### PR TITLE
Support styling on back button on iOS

### LINKED ISSUE #3454: [v2] `topBar.buttonColor` is ignored for back button
```
topBar: {
  buttonColor: 'red'
}
```

Did not work, but 

```
backButton: {
  color: 'red'
}
```

Did

### LINKED ISSUE #3731: [V2][iOS] Cannot set back button color
### Issue Description

When changing color of the back button of the `topBar` you define the styles like so:
```
Navigation.setDefaultOptions({
  topBar: {
    backButton: { color: '#ffffff' }, // Android
    buttonColor: '#ffffff', // iOS
  },
});
```

This works up to RNN version `2.0.2461`. But version `2.0.2462` onwards back button color is broken on iOS.

Looks like the only way to color your back button is to provide `icon` like this:
```
Navigation.setDefaultOptions({
  topBar: {
    backButton: { color: '#ffffff', icon: require('./images/someIcon.png') },
  },
});
```

But this is not ideal solution for people who want to use the default icon.

I am pretty sure these commits are related: https://github.com/wix/react-native-navigation/commit/98ff69ec562108f6b1e627804c52655494012d21 https://github.com/wix/react-native-navigation/commit/7cf172a364f6dcc17728c3e9e875f964ba679eb2

Ping @guyca @yogevbd 

---
### Environment
* React Native Navigation version: 2.0.2462
* React Native version: 0.54.4
* Platform(s): iOS
* Device info: Real iPhone 6s

### PR DESCRIPTION
**Pull request by ujwal-setlur**
This resolves #3454 and resolves #3731 

We can now specify:

```js
  Navigation.setDefaultOptions({
    topBar: {
      backButton: {
        color: "red",
        fontFamily: "Nunito-SemiBold",
        fontSize: 18,
        title: "boo"
      }
    }
  });
```

### CODE REVIEW

**Comment by henrikra** [comment]
I have now tested this PR and the color of the back button changed indeed 👍

File: lib/ios/RNNBackButtonOptions.m [hunk 1]
```diff
@@ -11,11 +11,44 @@ - (void)applyOn:(UIViewController *)viewController {
...
+		if(self.title) {
+			title = self.title;
+		} else {
+			title = viewController.navigationItem.title;
+		}
+
+		UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithTitle:title
 																	 style:UIBarButtonItemStylePlain
 																	target:nil
 																	action:nil];
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: Maybe too many empty lines here?

File: lib/ios/RNNBackButtonOptions.m [hunk 2]
```diff
@@ -11,11 +11,44 @@ - (void)applyOn:(UIViewController *)viewController {
 		
 		UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithImage:image style:UIBarButtonItemStylePlain target:nil action:nil];
 		viewController.navigationItem.backBarButtonItem = backItem;
-	} else if (self.title) {
-		UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithTitle:self.title
+	} else {
+		NSString *title;
+
+		if(self.title) {
+			title = self.title;
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: You could use ternary operator here like this to make this shorter and more clear: `NSString *title = self.title ? self.title : viewController.navigationItem.title;`

File: lib/ios/RNNBackButtonOptions.m [hunk 3]
```diff
@@ -11,11 +11,44 @@ - (void)applyOn:(UIViewController *)viewController {
...
+
+		if (self.color) {
+			[backItem setTintColor:[RCTConvert UIColor:self.color]];
+		}
+
+		if(self.fontFamily || self.fontSize) {
+			NSNumber* fontSize = self.fontSize;
+			NSString* fontFamily = self.fontFamily;
+			NSMutableDictionary* textAttributes = [[NSMutableDictionary alloc] init];
+			UIFont *font = nil;
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: Same here you could make simple ternary so there is no need for this `nil`

File: lib/ios/RNNBackButtonOptions.m [hunk 4]
```diff
@@ -11,11 +11,36 @@ - (void)applyOn:(UIViewController *)viewController {
 		
 		UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithImage:image style:UIBarButtonItemStylePlain target:nil action:nil];
 		[self setBackItem:backItem onViewController:viewController];
-	} else if (self.title) {
-		UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithTitle:self.title
+	} else {
+		NSString *title;
+		
+		title = self.title ? self.title : viewController.navigationItem.title;
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: Great! You can also combine this to one line `NSString *title = self.title ? self.title : viewController.navigationItem.title;`

File: lib/ios/RNNBackButtonOptions.m [hunk 5]
```diff
@@ -11,11 +11,36 @@ - (void)applyOn:(UIViewController *)viewController {
...
+
+		if (self.color) {
+			[backItem setTintColor:[RCTConvert UIColor:self.color]];
+		}
+
+		if(self.fontFamily || self.fontSize) {
+			NSNumber* fontSize = self.fontSize;
+			NSString* fontFamily = self.fontFamily;
+			NSMutableDictionary* textAttributes = [[NSMutableDictionary alloc] init];
+			UIFont *font;
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: Same thing here

File: lib/ios/RNNBackButtonOptions.m [hunk 6]
```diff
@@ -11,11 +11,36 @@ - (void)applyOn:(UIViewController *)viewController {
...
 																	 style:UIBarButtonItemStylePlain
 																	target:nil
 																	action:nil];
+
+		if (self.color) {
+			[backItem setTintColor:[RCTConvert UIColor:self.color]];
+		}
+
+		if(self.fontFamily || self.fontSize) {
+			NSNumber* fontSize = self.fontSize;
```
Comment (henrikra) [inline lib/ios/RNNBackButtonOptions.m]: Also here: `NSNumber* fontSize = self.fontSize ? self.fontSize : [[NSNumber alloc] initWithInt: 18];`

**Comment by khanh19934** [comment]
Is it support custom icon and custom component ?

**Comment by onlaps** [comment]
No, its to change font, title, color in ios version.

File: lib/ios/RNNBackButtonOptions.m (hunk 6 above)
Comment (ujwal-setlur) [reply lib/ios/RNNBackButtonOptions.m]: I prefer this for better readability frankly. This works.

**Comment by ujwal-setlur** [comment]
@henrika didn't know I had to add resolves in front of both issues. Good to know!

**Comment by rgoldiez** [comment]
@henrikra @ujwal-setlur - This PR works great.  Any idea on when it will be merged?

**Comment by aspidvip** [comment]
```
backButton: {
                 title: '',
                 color: 'white',
               },
```
not work

**Comment by abury** [comment]
Do we know when this is going to be merged? We really need a fix for this

**Comment by ujwal-setlur** [comment]
@aspidvip It works for me just fine:

```js
    topBar: {
      backButton: {
        color: "red",
        fontFamily: "Nunito-SemiBold"
      },
      title: {
        color: "red",
        fontFamily: "Nunito-Bold",
        fontSize: 18
      }
    }
```

**Comment by abury** [comment]
@ujwal-setlur What version are you on? That's not working as of v2.0.2425

**Comment by ujwal-setlur** [comment]
@abury I  tested this just yesterday with whatever is the latest version

**Comment by abury** [comment]
@ujwal-setlur The latest version as of writing on NPM is `2.0.2497`, and this fix isn't working.

**Comment by ujwal-setlur** [comment]
@abury I set color and font in default, and title when pushing screen. I will test it later tonight.

**Comment by henrikra** [comment]
@abury This PR hasn't meen merged so this fix cannot work 😆

**Comment by ujwal-setlur** [comment]
I thought of asking whether he had applied the patch, but then assumed he had 😊

**Comment by abury** [comment]
@henrikra I'm well aware the PR hasn't been merged, which is why I asked above when it will be. @ujwal-setlur commented directly after saying that a specific bit of code worked, so I thought it might not be needed, or was resolved in a separate PR. 👍

**Comment by ujwal-setlur** [comment]
@abury this PR is still needed and needs to be merged. For now I use patch-package to apply this patch every time I install react-native-navigation.

**Comment by heroic** [comment]
@guyca Any update on this?

**Comment by ujwal-setlur** [comment]
For some reason, this PR is not getting merged. People seem to want this. I certainly do. Is the design wrong? Is there a better way of doing this? Has this already been done in a different way?

This is an open source project, hopefully community contributions will be accepted.

**Comment by abury** [comment]
Agree with @ujwal-setlur. Not being able to change the color of the back button should be treated as a critical bug, not sure why this isn't being merged?
I certainly appreciate that this is open source and that people should appreciate the work done already (I certainly do), but it should also be appreciated that a lot of teams are deciding to put their faith behind this library and Wix over other options. I think it's reasonable to expect this type of issue to be addressed, especially if a PR has been submitted by the community.

File: lib/ios/RNNBackButtonOptions.m (hunk 4 above)
Comment (chrise86) [reply lib/ios/RNNBackButtonOptions.m]: I think this is wrong - it sets the back button title to the value of the current screen's title.

**Comment by troianoandres** [comment]
Any updates on this PR? This is something really critical since you are not able to change the color of the back button icon :/

**Comment by rgoldiez** [comment]
@yogevbd - Any plans on merging this PR?

**Comment by pie6k** [comment]
I've implemented custom navigation bar only because it's not possible to style back button :)

**Comment by Olovorr** [comment]
@guyca Could someone have a look at this PR please? 👼

**Comment by ujwal-setlur** [comment]
Thank you!

**Comment by henrikra** [comment]
Because there wasn’t any tests 🤷‍♂️

**Comment by chrise86** [comment]
Maybe, but that’s why we have pull reviews, right? Merging this has now created a new issue which will inevitable lead to many new issues reported here.

**Comment by ujwal-setlur** [comment]
@chrise,8, would you be happier if I set the title to blank if no title is set? That would be closer to original behavior. Also, you can always propose a change.

**Comment by ujwal-setlur** [comment]
If  @guyca reverted, please propose  a different fix. Clearly this functionality is needed. I really feel better coordination and cooperation is needed between core contributors and users that want to improve this.

**Comment by chrise86** [comment]
@ujwal-setlur I think you’ve misunderstood my intentions. I am very grateful to anyone who contributes to open source. I pointed out the issue with this PR because it needed testing and reviewing. I discovered the issue and reported back so that it wouldn’t get merged in its current state and cause more issues.

I have been testing my own solution to this problem too, but currently have nothing better. I do think a sensible solution would be to drop the title and just have the back arrow, yes.

**Comment by chrise86** [comment]
Once again I would like to thank you for your contribution and help driving this issue forward. But I don’t think, as it is, it is sensible to merge it in as people are only going to report the newly created issue. I hope you understand where I am coming from 😊

**Comment by ujwal-setlur** [comment]
@chrise86, no issues. I can make the change. I can't claim that my solutions are the best. I am more frustrated that that this project is not as communicative/responsive as others that I contribute to.

**Comment by chrise86** [comment]
@ujwal-setlur I understand your frustration, but I’m sure you can appreciate everyone is very busy. They’ve done a brilliant job getting the project to where it is, and with 265 open issues and 24 pull requests it’s going to take some time to get through it all.

I’m sure if we all work together we can help them out and make it even better 😊

File: lib/ios/RNNBackButtonOptions.m (hunk 4 above)
Comment (henrikra) [reply lib/ios/RNNBackButtonOptions.m]: @chrise86  So basically solution is to always use `self.title`?

Comment (chrise86) [reply lib/ios/RNNBackButtonOptions.m]: @henrikra kinda... what that does mean though, is that you will always have to pass a title to the back button in order. If you don't, then you will just get the back arrow alone.

Comment (henrikra) [reply lib/ios/RNNBackButtonOptions.m]: Please can you make PR out of this because this one died I guess :D

**Comment by henrikra** [comment]
I did bit more simple PR here https://github.com/wix/react-native-navigation/pull/3999

**Pull request merged**